In [1]:
#2018 HMDA Edit Testing File Generator
from collections import OrderedDict
import json
import pandas as pd
import random

import lar_constraints
import lar_generator



In [2]:
LAR_field_names = [
    "record_id","lei","uli","app_date","loan_type","loan_purpose",
    "preapproval","const_method","occ_type","loan_amount","action_taken",
    "action_date","street_address","city","state","zip_code","county",
    "tract","app_eth_1","app_eth_2","app_eth_3","app_eth_4","app_eth_5",
    "app_eth_code_14","app_eth_text","co_app_eth_1","co_app_eth_2",
    "co_app_eth_3","co_app_eth_4","co_app_eth_5","co_app_eth_code_14",
    "app_eth_basis","co_app_eth_basis","app_race_1","app_race_2",
    "app_race_3","app_race_4","app_race_5","app_race_code_1",
    "app_race_code_27","app_race_code_44","co_app_race_1","co_app_race_2",
    "co_app_race_3","co_app_race_4","co_app_race_5","co_app_race_code_1",
    "co_app_race_code_27","co_app_race_code_44","app_race_basis",
    "co_app_race_basis","app_sex","co_app_sex","app_sex_basis",
    "co_app_sex_basis","app_age","co_app_age","income","purchaser_type",
    "rate_spread","hoepa","lien","app_credit_score","co_app_credit_score",
    "app_score_name","app_score_code_8","co_app_score_name",
    "co_app_score_code_8","denial_1","denial_2","denial_3","denial_4",
    "denial_code_9","loan_costs","points_fees","origination_fee",
    "discount_points","lender_credits","interest_rate","prepayment_penalty",
    "dti","cltv","loan_term","intro_rate","balloon","int_only_pmts",
    "neg_amort","non_amort_features","property_value","manufactured_type",
    "manufactured_interest","total_units","affordable_units",
    "submission_type","initially_payable","mlo_id","aus_1","aus_2",
    "aus_3","aus_4","aus_5","aus_code_5","aus_result_1","aus_result_2",
    "aus_result_3","aus_result_4","aus_result_5","aus_code_16",
    "reverse_mortgage","open_end_credit","business_purpose"]

TS_field_names= [
    "record_id","inst_name","calendar_year","calendar_quarter",
    "contact_name","contact_tel","contact_email","contact_street_address",
    "office_city","office_state","office_zip","federal_agency",
    "lar_entries","tax_id","lei"]

In [3]:
#data lists used in the construction of LARs


In [4]:
#2018 Filing Instruction Guide: https://www.consumerfinance.gov/data-research/hmda/static/for-filers/2018/2018-HMDA-FIG.pdf
#Note: Add a format field? use to check format of entered data?

#Note: add valid value list/function to the valid vals list field
#IE list of ages 1 to n (range(200))
#IE rate_spread float generator



In [5]:

lar_schema_df = pd.DataFrame(json.load(open("../schemas/lar_schema.json", "r")))
ts_schema_df = pd.DataFrame(json.load(open("../schemas/ts_schema.json", "r")))

lar_schema_df.head()
ts_schema_df.head()

data_type   dtypes_             field valid_vals
0       numeric     [int]         record_id        [1]
1  alphanumeric  [string]         inst_name         []
2       numeric     [int]     calendar_year         []
3       numeric     [int]  calendar_quarter        [4]
4  alphanumeric  [string]      contact_name         []

In [10]:

boop = lar_generator.lar_gen(lar_schema_df, ts_schema_df) #instantiate generator
darp = lar_constraints.lar_constraints()#instantiate constraints

#test = lar_generator.lar_gen(lar_schema_df, ts_schema_df)
row = boop.make_row()
#set test values

#print(row_test["loan_purpose"], row_test["preapproval"], "purpose, pre")
#print(row["preapproval"],row["action_taken"])
row["city"] = ""; row['zip_code'] = ""; row["state"] = ""
print(row["street_address"], row["city"], row["zip_code"], row["state"])
#print(row["affordable_units"],":units", row["loan_purpose"],"purp")
#test functions in loop
#for func in darp.constraint_funcs:
#    row = getattr(darp, func)(row)
row = getattr(darp, "v622_const")(row)
#check modified output
print(row["street_address"], row["city"], row["zip_code"], row["state"])

#print("purpose:", row_test["loan_purpose"], "pre:", row_test["preapproval"])
#print("date:",row_test["app_date"], "action:",row_test["action_taken"])



1234 Hocus Potato Way   
1234 Hocus Potato Way Spudfarm 12345 UT


In [ ]:
#Quality and Macro field interrelationship constraints:

#set file length parameters
#LAR rows (small=200, medium=1,000, large=10,000)
lar_small = 200
lar_medium = 1000
lar_large = 10000

#Notes: increase chance for enum added to range
# free text not valid without certain selections
# mlo_id needs NA option




#set dummy values for TS row
ts_row_small = OrderedDict()

ts_row_small["record_id"]="1"
ts_row_small["inst_name"]="Ficus Bank"
ts_row_small["calendar_year"]=str(2018)
ts_row_small["calendar_quarter"]=str(TS_df.valid_vals[TS_df.field=="calendar_quarter"].iloc[0][0])
ts_row_small["contact_name"]="Mr. Smug Pockets"
ts_row_small["contact_tel"]="555-555-5555"
ts_row_small["contact_email"]="pockets@ficus.com"
ts_row_small["contact_street_address"]="1234 Ficus Lane"
ts_row_small["office_city"]="Ficusville"
ts_row_small["office_state"]="UT"
ts_row_small["office_zip"]="84096"
ts_row_small["federal_agency"]=str(TS_df.valid_vals[TS_df.field=="federal_agency"].iloc[0][0])
ts_row_small["lar_entries"]=str(lar_small)
ts_row_small["tax_id"]="01-0123456"
ts_row_small["lei"]="12345678901234567890"

#create TS medium row
ts_row_medium = ts_row_small.copy()
ts_row_medium["lar_entries"] = str(lar_medium)

#create TS large row
ts_row_large = ts_row_small.copy()
ts_row_large["lar_entries"] = str(lar_large)

